desc/compute/_stability.py

```diff
@@ -234,199 +235,210 @@ def _magnetic_well(params, transforms, profiles, data, **kwargs):
-    a_N = data["a"]
-    B_N = 2 * psi_b / a_N**2
-
-    zeta0 = kwargs.get("zeta0", jnp.linspace(-0.5 * jnp.pi, 0.5 * jnp.pi, 15))
-    N_zeta0 = len(zeta0)
-
-    # This would fail with rho vectorization
-    iota = jnp.mean(data["iota"])
-    shear = jnp.mean(data["shear"])
-    psi = jnp.mean(data["psi"])
-    sign_psi = jnp.sign(psi)
-    sign_iota = jnp.sign(iota)
-
-    N_rho = int(source_grid.num_rho)
-    N_alpha = int(source_grid.num_alpha)
-
-    # phi is the same for each alpha
-    phi = source_grid.nodes[:: N_rho * N_alpha, 2]
-    N_zeta = len(phi)
-
-    B = source_grid.meshgrid_reshape(data["|B|"], "arz")
-    B_sup_zeta = source_grid.meshgrid_reshape(data["B^zeta"], "arz")
-    gradpar = B_sup_zeta / B
-
-    # This would fail with rho vectorization
-    dpdpsi = jnp.mean(mu_0 * data["p_r"] / data["psi_r"])
-
-    g_sup_aa = source_grid.meshgrid_reshape(data["g^aa"], "arz")[None, ...]
-    g_sup_ra = source_grid.meshgrid_reshape(data["g^ra"], "arz")[None, ...]
-    g_sup_rr = source_grid.meshgrid_reshape(data["g^rr"], "arz")[None, ...]
-
-    gds2 = jnp.reshape(
-        jnp.transpose(
-            rho**2
-            * (
-                g_sup_aa
-                - 2 * sign_iota * shear / rho * zeta0[:, None, None, None] * g_sup_ra
-                + zeta0[:, None, None, None] ** 2 * (shear / rho) ** 2 * g_sup_rr
-            ),
-            axes=(1, 0, 2, 3),
-        ),
-        (N_alpha, N_zeta0, N_zeta),
-    )
-
-    f = a_N * B_N**3 * gds2 / B**3 * 1 / gradpar
-    g = a_N**3 * B_N * gds2 / B * gradpar
-    g_half = (g[:, :, 1:] + g[:, :, :-1]) / 2
-
-    cvdrift = source_grid.meshgrid_reshape(data["cvdrift"], "arz")[None, ...]
-    cvdrift0 = source_grid.meshgrid_reshape(data["cvdrift0"], "arz")[None, ...]
-
-    c = (
-        a_N**3
-        * B_N
-        * jnp.reshape(
-            jnp.transpose(
-                2
-                / B_sup_zeta[None, ...]
-                * sign_psi
-                * rho**2
-                * dpdpsi
-                * (
-                    cvdrift
-                    - shear / (2 * rho**2) * zeta0[:, None, None, None] * cvdrift0
-                ),
-                axes=(1, 0, 2, 3),
-            ),
-            (N_alpha, N_zeta0, N_zeta),
+    eigfuns = kwargs.get("eigfuns", True)
+    grid = transforms["grid"].source_grid
+    # toroidal step size between points along field lines is assumed uniform
+    dz = grid.nodes[grid.unique_zeta_idx[:2], 2]
+    dz = dz[1] - dz[0]
+    num_zeta0 = data["c balloon"].shape[0]
+
+    def reshape(f):
+        assert f.shape == (num_zeta0, grid.num_nodes)
+        f = jnp.swapaxes(grid.meshgrid_reshape(f.T, "raz"), -1, -2)
+        assert f.shape == (grid.num_rho, grid.num_alpha, num_zeta0, grid.num_zeta)
+        return f
+
+    c = reshape(data["c balloon"])
+    f = reshape(data["f balloon"])
+    g = reshape(data["g balloon"])
+    # Approximate derivative along field line with second order finite differencing.
```

Comment:
we could do chain rule like i did for the bounce integral stuff to get spectrally accurate derivatives but its fine